src/test/java/com/amazon/ion/util/ZstdStreamInterceptorTest.java

```diff
@@ -0,0 +1,379 @@
+        IonReaderBuilder builder = ((_Private_IonReaderBuilder) IonReaderBuilder.standard())
+            .withCustomClassLoader(new CustomInterceptorClassLoader());
+        assertGzipThen(builder, ZstdStreamInterceptor.class);
+    }
+
+    private static class DummyInterceptor implements InputStreamInterceptor {
+
+        @Override
+        public String formatName() {
+            return null;
+        }
+
+        @Override
+        public int headerMatchLength() {
+            return 0;
+        }
+
+        @Override
+        public boolean matchesHeader(byte[] candidate, int offset, int length) {
+            return false;
+        }
+
+        @Override
+        public InputStream newInputStream(InputStream interceptedStream) {
+            return null;
+        }
+    }
+
+    @Test
+    public void addingManuallyTakesPrecedenceOverClasspath() {
+        IonReaderBuilder builder = ((_Private_IonReaderBuilder) IonReaderBuilder.standard())
+            .withCustomClassLoader(new CustomInterceptorClassLoader()) // This would add Zstd if classpath detection were used.
+            .addInputStreamInterceptor(new DummyInterceptor());
+        assertGzipThen(builder, DummyInterceptor.class);
+    }
+
+    @Test
+    public void differentClassLoadersInEachThread() throws Exception {
+        Thread withDefaultClassLoader = new Thread(() -> {
+            IonReaderBuilder readerBuilder = IonReaderBuilder.standard();
+            List<InputStreamInterceptor> streamInterceptors = readerBuilder.getInputStreamInterceptors();
+            assertEquals(1, streamInterceptors.size());
+            assertSame(GZIPStreamInterceptor.INSTANCE, streamInterceptors.get(0));
+        });
+        Thread withCustomClassLoader = new Thread(() -> {
+            IonReaderBuilder readerBuilder = IonReaderBuilder.standard();
+            List<InputStreamInterceptor> streamInterceptors = readerBuilder.getInputStreamInterceptors();
+            assertGzipThen(readerBuilder, ZstdStreamInterceptor.class);
+            // Verify that a new IonReaderBuilder instance does not re-detect the interceptors applicable to this
+            // thread.
+            readerBuilder = IonReaderBuilder.standard();
+            assertSame(streamInterceptors, readerBuilder.getInputStreamInterceptors());
+        });
+        withCustomClassLoader.setContextClassLoader(new CustomInterceptorClassLoader());
+
+        withDefaultClassLoader.start();
+        withCustomClassLoader.start();
+
+        // While the spawned threads are working, verify they do not affect the parent thread.
+        IonReaderBuilder builder = IonReaderBuilder.standard().addInputStreamInterceptor(new DummyInterceptor());
+        assertGzipThen(builder, DummyInterceptor.class);
+
+        withDefaultClassLoader.join();
+        withCustomClassLoader.join();
+    }
+
+    private static class LengthTooLongInterceptor implements InputStreamInterceptor {
```

Comment:
A shared stub implementation can reduce repetition in interceptor implementation in this test.